### PR TITLE
[SPARK-8432] [SQL] fix hashCode() and equals() of BinaryType in Row

### PR DESCRIPTION
**Pull request by davies**
Also added more tests in LiteralExpressionSuite

### CODE REVIEW

File: sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java [hunk 1]
```diff
@@ -215,4 +227,15 @@ public String mkString(String sep) {
...
+
+  /*
+   * Returns hash code based on bytes in `arr`
+   * */
+  protected int bytesHashCode(byte[] arr) {
+    int hash = 0;
+    for (int i = 0; i < arr.length; i++) {
+      hash = hash * 37 + (int)arr[i];
+    }
+    return hash;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java]: Is it same with `java.util.Arrays.hashCode`?

File: sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java [hunk 2]
```diff
@@ -167,9 +167,21 @@ public boolean equals(Object other) {
...
+            // handle equals() of byte[]
+            if (!(o2 instanceof byte[]) || !java.util.Arrays.equals((byte[])o1, (byte[])o2)) {
+              return false;
+            }
+          } else if (!o1.equals(o2)) {
+            return false;
+          }
+        }
       }
       return true;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java]: We defined `equals` at `Row`, and override it at `BaseRow` and `GenericRow`. But these 3 `equals` are very similar, can we combine them into one?

File: sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java (hunk 1 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java]: Good idea, we should use that.

File: sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java]: Because `Row` is a trait, so we can't use it in Java (for UnsafeRow and SpecificRow in codegen), so we at least need two version of it.

**Comment by davies** [comment]
@marmbrus Could you help to review this one?

File: sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java [hunk 3]
```diff
@@ -155,27 +155,6 @@ public int fieldIndex(String name) {
     throw new UnsupportedOperationException();
   }
 
-  /**
-   * A generic version of Row.equals(Row), which is used for tests.
-   */
-  @Override
```
Comment (marmbrus) [inline sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java]: Existing: can you add some javadoc to this class to explain what its used for and why its in Java?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 4]
```diff
@@ -27,6 +27,63 @@ import org.apache.spark.sql.catalyst.expressions.GenericRow
 abstract class InternalRow extends Row {
   // A default implementation to change the return type
   override def copy(): InternalRow = {this}
+
+  // A default version (slow), used for tests
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: Is this really only used for tests?  Won't it be used when we use an internal row as a HashKey?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala [hunk 5]
```diff
@@ -55,7 +55,7 @@ trait ExpressionEvalHelper {
     val actual = try evaluate(expression, inputRow) catch {
       case e: Exception => fail(s"Exception evaluating $expression", e)
     }
-    if (actual != expected) {
+    if (actual !== expected) {
```
Comment (marmbrus) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: What are the semantics of `!==`?  This is new to me.

**Comment by marmbrus** [comment]
@davies, thanks for working on this!  I'm okay with this approach, but did you consider the alternative, where we instead change the internal type of `BytesType` to be a wrapper class instead of `Array[Byte]`, and define a good equality function for that wrapper?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala (hunk 4 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: Yes, you are right. The comment is only true for codegen, which has a special version of `equals`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala (hunk 5 above)
Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala]: Just learned from someone else, thought it could help to compare Array, will revert it.

File: sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java (hunk 3 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/BaseRow.java]: Because `Row` is a trait, UnsafeRow and SpecificRow are both in Java, they can not inherit some default implementations from `Row`, so created BaseRow in Java for them. Right now, we have `InternalRow`, will be clean these in another PR.

**Comment by davies** [comment]
@marmbrus We're working to have more efficient representation in catalyst, putting Array[Byte] inside a wrapper sounds not in the same direction. I'd like to go this approach.

**Comment by marmbrus** [comment]
I think using a wrapper might be necessary for efficiency. For example, we will want to reuse the same byte array when reading from something like parquet, instead of needing to allocate one of the exact size each time (think `java.nio.ByteBuffer`).  However, I'm fine merging this first.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 6]
```diff
@@ -26,7 +26,67 @@ import org.apache.spark.sql.catalyst.expressions.GenericRow
  */
 abstract class InternalRow extends Row {
   // A default implementation to change the return type
-  override def copy(): InternalRow = {this}
+  override def copy(): InternalRow = this
+
+  override def equals(o: Any): Boolean = {
+    if (!o.isInstanceOf[Row]) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: Will we change it to `isInstanceOf[InternalRow]` after https://github.com/apache/spark/pull/6869?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 7]
```diff
@@ -26,7 +26,67 @@ import org.apache.spark.sql.catalyst.expressions.GenericRow
...
+      return false
+    }
+
+    val other = o.asInstanceOf[Row]
+    if (length != other.length) {
+      return false
+    }
+
+    for (i <- 0 until length) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: I think we should use `while` here, as `for` is inefficient in scala.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala (hunk 6 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: yes

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala [hunk 8]
```diff
@@ -26,7 +26,67 @@ import org.apache.spark.sql.catalyst.expressions.GenericRow
...
+        if (o1.isInstanceOf[Array[Byte]]) {
+          // handle equality of Array[Byte]
+          val b1 = o1.asInstanceOf[Array[Byte]]
+          if (!o2.isInstanceOf[Array[Byte]] ||
+            !java.util.Arrays.equals(b1, o2.asInstanceOf[Array[Byte]])) {
+            return false
+          }
+        } else if (o1 != o2) {
+          return false
+        }
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: what about
```scala
(o1, o2) match {
  case (b1: Array[Byte], b2: Array[Byte]) => java.util.Arrays.equals(b1, b2)
  case _ => o1 == o2
}
```

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: `match` is also slow in Scala

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralExpressionSuite.scala [hunk 9]
```diff
@@ -31,8 +45,16 @@ class LiteralExpressionSuite extends SparkFunSuite with ExpressionEvalHelper {
   }
 
   test("int literals") {
-    checkEvaluation(Literal(1), 1)
-    checkEvaluation(Literal(0L), 0L)
+    List(0, 1, Int.MinValue, Int.MaxValue).foreach {
```
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralExpressionSuite.scala]: this is a pretty weird way of indenting. you can do

```scala
List(0, 1, Int.MinValue, Int.MaxValue).foreach { d =>
  ...
}
```

or
```scala
for (d <- List(0, 1, Int.MinValue, Int.MaxValue)) {
  ...
}
```

**Comment by marmbrus** [comment]
test this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala [hunk 10]
```diff
@@ -127,6 +127,7 @@ object GenerateProjection extends CodeGenerator[Seq[Expression], Projection] {
         case FloatType => s"Float.floatToIntBits($col)"
         case DoubleType =>
             s"(int)(Double.doubleToLongBits($col) ^ (Double.doubleToLongBits($col) >>> 32))"
+        case BinaryType => s"java.util.Arrays.hashCode($col)"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala]: Should we also update `equals` for generated code?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala (hunk 8 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/InternalRow.scala]: Should we abstract this column compare logic into a method? So that we don't need to write it again for code generation...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala (hunk 10 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala]: That's already done

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateProjection.scala]: Ah I see, `genEqual` has already handled `BinaryType`.

**Comment by marmbrus** [comment]
Thanks, merging to master.
